### DISCUSSION TITLE
Logback has removed or.slf4j.impl package from logback-classic 1.3.x onwards

### DISCUSSION THREAD

**Discussion by skumarjh-tibco** [qos-ch/logback#793] [General]
Hello experts, 
I wonder why the package "org.slf4j.impl" package has been removed from the logback-classic 1.3.x version. 
It is available till the last version of 1.2.x.
I can see the same package with the same classes is available in the log4j-slf4j-impl jar.
When I use this jar to use the org.slf4j.impl in my project, I get no class def found for threadcontext.
When I add the log4j-slfj-impl to the classpath, i am getting "failed. class org.apache.logging.slf4j.Log4jLoggerFactory cannot be cast to class ch.qos.logback.classic.LoggerContext"
Can someone help me if I am using the correct jar or can someone help me to find if logback-classic's org.slf4j.impl package is bundled with any other specific jar?
This will help me a lot.

Thanks in advance.

**Comment by ceki** [comment]
@skumarjh Logback version 1.3.x and later require SLF4J 2.0.x or later. As such, `org.slf4j.impl` package is no longer included. See also [what changed in SLF4J 2.0](https://www.slf4j.org/faq.html#changesInVersion200)

**Reply by skumarjh-tibco** [reply]
Thank you for your response @ceki. 
I have gone through the details you shared about what changed with the logback. 
I am still in doubt and have a question about the classes inside the particular package "org.slf4j.impl". How can I get those classes if I require it and have used inside my projects. I have multiple dependencies on this package and If I remove this dependency, I am getting the dependency error. 
Can you help me with this? Is it completely removed now or the classes inside the package are added to any other jar or package?
This will help me a lot to proceed further.
Thank you again.

**Reply by ceki** [reply]
Can you share at least one of the error messages? Which logging libraries are you using?

**Reply by skumarjh-tibco** [reply]
Hi @ceki , we are getting below error. This happened while we upgraded the logback jars to 1.4.x latest. 
While upgrading, we noticed that the package "org.slf4j.impl" is missing from logback classic which we have used at multiple places. To get that, we found onejar "log4j-slf4j-impl" which contained the same package and we added that jar in our classpath. 
As I read and found that the package "org.slf4j.impl" is no more available in classic, I could not find any other way to resolve that dependency. But, with the jar "log4j-slf4j-impl", we noticed that this is implementation of log4j but we are using slf4j. 

Below is the error which are getting after these upgrade.
 [exec] Exception in thread "main" java.lang.NoClassDefFoundError: org/apache/logging/log4j/ThreadContext
     [exec] 	at org.apache.logging.slf4j.Log4jMDCAdapter.put(Log4jMDCAdapter.java:30)
     [exec] 	at org.slf4j.MDC.put(MDC.java:147)
     --------------------------------------
[exec] Caused by: java.lang.ClassNotFoundException: org.apache.logging.log4j.ThreadContext
     [exec] 	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
     [exec] 	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
     [exec] 	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
     [exec] 	... 4 more
     
We also encountered one error at runtime which is similar to https://github.com/qos-ch/logback/issues/723
     Can you help me with if the package "org.slf4j.impl" is no more available to use or is there any alternative for it?
     This would help me to proceed further. 

 Thanks.

**Comment by ceki** [comment]
@skumarjh 
Regarding the error above, `log4j-slf4j-impl` requires log4j2-api.jar (and probably also log4j2-core.jar) which are apparently missing from your classpath.

Which version or versions of slf4j do you have on your classpath? Each copy of  `org.sff4j.impl` found in various logging implementation is different. These copies are **not** interchangeable.  As such, trying to place `org.sff4j.impl` package on your classpath makes no sense. In my opinion, you should stop trying to find `org.slf4j.impl` from a random sources.

**Reply by skumarjh-tibco** [reply]
Thank you for quick response @ceki .
I am using slf4j 2.0.9. 
I have used log4j-slf4j-impl only because I could not find the package "org.slf4j.impl" anywhere else. 
Lets take it this way, if we do not use log4j-slf4j-impl jar or any other extra jar, what should we do if we need that older package "org.slf4j.impl" which was in logback-classic till 1.2.x?
What do you suggest on this?

Thanks.

**Reply by ceki** [reply]
"org.slf4j.impl" was required by slf4j 1.7 and earlier to bind/link with the logging back-end. With slf4j 2.x the "org.slf4j.impl" is no longer used nor needed. If  you are using slf4j 2.x and logback 1.3+, you don't need it either.

**Reply by skumarjh-tibco** [reply]
Thanks for your response @ceki 
We have been using MDC in our projects which indirectly uses threadcontext which is part of org.slf4j.impl.
With the upgrade, we are having issues with this as, logback doesn't have the org.slf4j.impl which is needed. 
We have also tried the same by removing all dependencies on log4j but still resulted in error.
If this no longer part of logback, can you please suggest any alternatives?
Thanks

**Comment by ceki** [comment]
@skumarjh You can use `org.slf4j.MDC` without the `org.slf4j.impl` package. Moreover, `threadcontext` is **not** part of `org.slf4j.impl`.

**Reply by skumarjh-tibco** [reply]
Hi @ceki , We have tried multiple things at our end in order to resolve the issue but still we are facing some issues. 
I am simplifying my question here, 
If the particular package "org.slf4j.impl" has been deleted and there is not alternative for it, but we have used that package at multiple places in our project, what should we do?
Should we completely remove it from our projects or is there anything else we can do ?

Thanks.

**Comment by ceki** [comment]
@skumarjh https://github.com/qos-ch/slf4j/issues/415#issuecomment-2061741015 might be related.
